abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py

```diff
@@ -36,3 +37,38 @@ def F_reloc_LR(vel_now, V_now, v_desired=None, theta_max=None):
     D_leftright = left_excitation - right_excitation
     theta = D_leftright * theta_max
     return (v_desired - vel_now), theta
+
+
+def reflection_from_circular_wall(dx, dy, orientation):
+    """
+    Calculating the reflection of the agent from the circle arena border.
+    SEE: https://stackoverflow.com/questions/54543170/angle-reflexion-for-bouncing-ball-in-a-circle
+
+    :param dx: x coordinate of the agent minus center of the circle
+    :param dy: y coordinate of the agent minus center of the circle
+    :param orientation: orientation of the agent
+    :return: new orientation of the agent
+    """
+    # normal vector of the circle
+    c_norm = (pygame.math.Vector2(dx, dy)).normalize()
+    # incident vector: the current direction vector of the bouncing agent
+    vec_i = pygame.math.Vector2(np.cos(orientation), np.sin(orientation))
+    # orientation inside the circle
+    i_orientation = np.pi + np.arctan2(vec_i[1], vec_i[0])
+
+    # reflection vector: outgoing direction vector of the bouncing agent
+    vec_r = vec_i - 2 * c_norm.dot(vec_i) * c_norm
+    # np.degrees(self.orientation)
+    new_orientation = np.pi + np.arctan2(vec_r[1], vec_r[0])
+
+    # make sure that the new orientation points inside the circle and not too
+    # flat to the border
+    if np.abs(new_orientation - i_orientation) > np.pi / 4:
```

Comment:
I understand this fixes the behavior, but I am wondering why it is necessary.

Reply:
I would guess it has something to do with the curvature of the boundary and the step size per update 😅